receiver/prometheusreceiver: add option to fallback to collector starttime

ridwanmsharif opened the pull request:
#### Description
This change adds an option to the metric adjuster to use an approximation of the collector starttime as a fallback for the start time of scraped cumulative metrics. This is useful when no start time is found and when the collector starts up alongside its targets (like in serverless environments or sidecar approaches).

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/36364


<!--Describe what testing was performed and which tests were added.-->
#### Testing
Added unit test for this config option

<!--Describe the documentation added.-->
#### Documentation
Config option added to the README.

Commit:
receiver/prometheusreceiver: add option to fallback to collector starttime

This change adds an option to the metric adjuster to use an
approximation of the collector starttime as a fallback for the start
time of scraped cumulative metrics. This is useful when no start time is
found and when the collector starts up alongside its targets (like in
serverless environments or sidecar approaches).

Signed-off-by: Ridwan Sharif <ridwanmsharif@google.com>

ArthurSens commented:
The code itself looks correct! 

To be completely honest, I'm pretty new to this component and haven't used it myself. I noticed we have waaaay too many fallbacks for Created Timestamps; that looks weird 🤔.

Do I understand correctly that the flow is like this:
* If metric `StartTimeUnixNano` is set, use that to populate the Created Timestamp from the prometheus/client_golang SDK.
* If `StartTimeUnixNano` is not set, get it from another metric called `process_start_time_seconds` (where does that come from?)
* Finally, if we still don't have a timestamp, we use the collector start time.

Did I get the flow correctly?


It makes me wonder, when would an OpenTelemetry metric not have `StartTimeUnixNano` not set? I understand this is not a required field in the spec, but maybe we could work on making it required instead?

ArthurSens commented on receiver/prometheusreceiver/README.md:
```diff
@@ -128,6 +128,7 @@ receivers:
       trim_metric_suffixes: true
       use_start_time_metric: true
       start_time_metric_regex: foo_bar_.*
+      use_collector_start_time_fallback: true
```
```suggestion
      use_collector_start_time_fallback: false
```

Should we give the default value as example? :P

dashpole replied:
we use a non-default for all other options, so this seems ok to me

dashpole commented on receiver/prometheusreceiver/internal/starttimemetricadjuster.go:
```diff
@@ -19,21 +20,32 @@ var (
+func NewStartTimeMetricAdjuster(logger *zap.Logger, startTimeMetricRegex *regexp.Regexp, useCollectorStartTimeFallback bool) MetricsAdjuster {
+	var fallbackStartTime *time.Time
+	if useCollectorStartTimeFallback {
+		now := time.Now()
```
Is that correct in a containerized environment, or would it give the start time of the host?

dehaansa replied:
I don't know the answer to that question, would need to be tested (I don't currently have capacity to test myself).

ridwanmsharif replied:
Am now using `github.com/shirou/gopsutil/v4/process` and it does give the correct process start time in containerized environments too. Thanks for the suggestion.

Aneurysm9 commented on receiver/prometheusreceiver/internal/starttimemetricadjuster.go:
```diff
@@ -19,21 +20,32 @@ var (
+func NewStartTimeMetricAdjuster(logger *zap.Logger, startTimeMetricRegex *regexp.Regexp, useCollectorStartTimeFallback bool) MetricsAdjuster {
+	var fallbackStartTime *time.Time
+	if useCollectorStartTimeFallback {
+		now := time.Now()
```
This won't be the collector start time, but the start time of this instance of the component.  If the pipeline is stopped and restarted then this will result in a different timestamp even though the process has not restarted.  This is a subtlety, but could be very confusing for someone using OpAMP to manage collector instances.  I'm not sure whether there's a way to get a reliable process start time from the collector host.  The system uptime is almost certainly not the correct value to use here.  Perhaps the best option is just to clarify in the description of the new configuration field that the approximated start time will be relative to the component start time and not necessarily the collector start time.

dashpole replied:
can we populate it using an init function, or a variable outside of metric adjuster?

Aneurysm9 replied:
Populating a variable in an init function could work.  It would be executed once near the start of the process.

ridwanmsharif replied:
Done.

Aneurysm9 commented on receiver/prometheusreceiver/internal/starttimemetricadjuster.go:
```diff
@@ -19,21 +20,32 @@ var (
+		if stma.fallbackStartTime == nil {
+			return err
+		}
+		stma.logger.Warn("Couldn't get start time for metrics. Using fallback start time.", zap.Error(err))
```
Does this need to be at the `Warn` level?  Wouldn't this be a fairly high-rate log entry if none of the processed metrics have a start time?

ridwanmsharif replied:
Good point, made it an info log instead.

Aneurysm9 commented:
> It makes me wonder, when would an OpenTelemetry metric not have `StartTimeUnixNano` not set? I understand this is not a required field in the spec, but maybe we could work on making it required instead?

This is the prometheus receiver, so the concern here is prometheus metrics not having a start time and we want to ensure that the `pdata` metrics produced by the receiver do have a start time.  The flow you described appears correct, though it is in the context of populating `StartTimeUnixNano` instead of looking to it for a value.

ridwanmsharif commented:
Sorry I let this get stale, I was ooo for the past couple months but am getting back to this today. I'll make the change to set the start time in an init function, that feels most appropriate here

Commit:
prometheusreceiver: approximate collector start time in an init function

Commit:
prometheusreceiver: use gopsutil to gather collector start time on component initialization

dashpole reviewed (approved):
Can you add a feature gate around this (alpha) so this is easier for us to remove later if we move it to the metric start time processor?

dashpole commented:
(accidental approval)

Commit:
prometheusreceiver: change start time fallback from a config knob to a featuregate

This change creates an Alpha feature gate for this functionality since
this can be replaced when
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/37186
is implemented.

ridwanmsharif commented:
> Can you add a feature gate around this (alpha) so this is easier for us to remove later if we move it to the metric start time processor?

Done. Makes sense to me

dehaansa commented on receiver/prometheusreceiver/internal/starttimemetricadjuster.go:
```diff
@@ -15,25 +18,55 @@ var (
+func init() {
+	p, err := process.NewProcess(int32(os.Getpid()))
+	if err != nil {
+		panic("Unable to find current process" + err.Error())
```
I'm a little concerned about being able to panic in code that supports a feature-gated feature. I understand these are very unlikely panics, but is that the right behavior for this error path?

ridwanmsharif replied:
After some playing around, and given that this is a featuregated feature - I think its simpler to just approximate the collector start time using time.Now() in an init function. No error paths this way and this is supposed to be an approximation anyway.

dashpole commented on .chloggen/starttime-fallback.yaml:
```diff
@@ -0,0 +1,27 @@
+component: prometheusreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Add `UseCollectorStartTimeFallback` featuregate for the start time metric adjuster to use the collector start time as an approximation of process start time as a fallback.
```
```suggestion
note: Add `receiver.prometheusreceiver.UseCollectorStartTimeFallback` featuregate for the start time metric adjuster to use the collector start time as an approximation of process start time as a fallback.
```

dashpole commented on receiver/prometheusreceiver/internal/starttimemetricadjuster.go:
```diff
@@ -15,25 +18,55 @@ var (
+		panic("Unable to find current process" + err.Error())
+	}
+
+	ct, err := p.CreateTime()
```
Can we just use `var collectorStartTime = time.Now()`?

dashpole commented on receiver/prometheusreceiver/internal/starttimemetricadjuster.go:
```diff
@@ -15,25 +18,55 @@ var (
 	startTime, err := stma.getStartTime(metrics)
 	if err != nil {
-		return err
+		if stma.fallbackStartTime == nil {
```
You can just evaluate the feature gate here, rather than plumbing booleans through the code base.

Commit:
prometheusreceiver: apply suggestions from review
